### ISSUE TITLE
Please allow for installing Python dependencies from requirements files using wheels

### ISSUE THREAD

**Issue by Vultaire** [canonical/charmcraft#712]
At present, packages from requirements files are installed via pip using the ```--no-binary :all:``` flag.

This can result in exceptionally long build times which, if the charm author is OK with using pre-built wheels, are unnecessarily long.

I'm working on a charm, and its requirements.txt looks like this:

```
charmhelpers
ops
click
fabric
python-openstackclient
```

Building the venv for the above requirements takes under 30 seconds on my machine.  (Tested via both Focal and Bionic LXDs, using ```time { python3 -m venv venv; . venv/bin/activate; pip install --upgrade setuptools pip; pip install --upgrade -r requirements.txt; }```.)

I would expect charmcraft.yaml to look something like this:

```
type: "charm"
bases:
  - name: "ubuntu"
    channel: "20.04"
    architectures:
      - amd64
  - name: "ubuntu"
    channel: "18.04"
    architectures:
      - amd64
parts:
  charm:
    charm-binary-python-packages:
      # Updated versions needed for working nicely with
      # jinja2/markupsafe
      - setuptools
      - pip
```

However, the packages in the requirements.txt depend on a wide range of other packages, and with the ```--no-binary :all:``` flag, it takes a very long time - longer than I want to clock for this bug.

As an attempt at working around this issue, I've customized my charmcraft.yaml to look like this instead:

```
parts:
  charm:
    build-packages:
      # Needed by cryptography
      - libffi-dev
      - rustc
      - cargo
      - libssl-dev
    charm-binary-python-packages:
      # Updated versions needed for working nicely with
      # jinja2/markupsafe
      - setuptools
      - pip

      # Needed as prerequisites for other required packages.
      # Explicitly included here to avoid doing source-based builds
      # which take time and require installing extra dev packages.
      - setuptools-rust
      #- cryptography   # note: seems to require a source build on Bionic,
                        # which in turn requires setuptools-rust to be installed
                        # before attempting to install this.
      - bcrypt
      - pynacl          # note: no binary on bionic, but this does help focal

      # Include any of the requirements.txt packages which have wheels
      #- charmhelpers   # currently tarball-only
      - ops
      - click
      #- fabric         # currently tarball-only
      #- python-openstackclient  # pulls cryptography dependency before
                                 # setuptools-rust can be installed
```

While this does install a few more bits via binaries when possible, my Bionic build time from venv creation through final dependency install (and excluding the other parts of the charm building process) takes roughly 13.5 minutes.  Again, doing this via a normal venv only takes me roughly 30 seconds since most of the packages can be installed via wheels.

Long story short: I don't see a good workaround at present to get reasonable build times for my particular required packages when using Charmcraft 1.4.0.  Can we add an option so we can omit the ```--no-binary :all:``` when installing from requirements.txt files, or something similar in nature?

Test environment:
Host: Ubuntu 20.04
Charmcraft: latest/stable (1.4.0)
Invoked via: ```charmcraft -v pack```
Building from bases: 20.04 and 18.04

**Comment by facundobatista** [comment]
> This can result in exceptionally long build times which, if the charm author is OK with using pre-built wheels, are unnecessarily long.

It can take a lot of time the first time it's built. But after that, all calls to `charmcraft pack` will reuse the previously built packages, so it will be fast. Or you're experiencing long times for building everytime? If yes, could you please provide logs when running `charmcraft --verbose pack`? Thanks!

Regarding building from binaries, there are no plans to allowing that easily or by default, as the recommended way is building from sources so all building environment is controlled.

**Comment by addyess** [comment]
The build times are very long when run in CI because those environments are ephemeral.  This same kind of thing continues to plague me with build times of 20 to 30 minutes where the `requirements.txt` isn't dramatic.

a single package example from [this ci run](https://github.com/charmed-kubernetes/vsphere-cloud-provider-operator/runs/6716382063?check_suite_focus=true)

```log
2022-06-02 21:26:50.033 :: 2022-06-02 21:26:50.008    :: Collecting jsonschema 
2022-06-02 21:26:50.048 :: 2022-06-02 21:26:50.033    ::   Downloading jsonschema-4.6.0.tar.gz (269 kB)                                                        
2022-06-02 21:26:50.277 :: 2022-06-02 21:26:50.048    ::      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 269.9/269.9 kB 24.3 MB/s eta 0:00:00                      
2022-06-02 21:28:30.497 :: 2022-06-02 21:26:50.277    ::   Installing build dependencies: started                                                              
2022-06-02 21:30:15.945 :: 2022-06-02 21:28:30.497    ::   Installing build dependencies: still running...                                                     
2022-06-02 21:32:01.929 :: 2022-06-02 21:30:15.945    ::   Installing build dependencies: still running...                                                     
2022-06-02 21:33:11.087 :: 2022-06-02 21:32:01.928    ::   Installing build dependencies: still running...                                                     
2022-06-02 21:34:39.126 :: 2022-06-02 21:33:11.086    ::   Installing build dependencies: still running...                                                     
2022-06-02 21:35:40.121 :: 2022-06-02 21:34:39.126    ::   Installing build dependencies: still running...                                                     
2022-06-02 21:35:50.309 :: 2022-06-02 21:35:40.120    ::   Installing build dependencies: still running...                                                     
2022-06-02 21:35:50.314 :: 2022-06-02 21:35:50.309    ::   Installing build dependencies: finished with status 'done'                                          
2022-06-02 21:35:50.382 :: 2022-06-02 21:35:50.313    ::   Getting requirements to build wheel: started                                                        
2022-06-02 21:35:50.383 :: 2022-06-02 21:35:50.382    ::   Getting requirements to build wheel: finished with status 'done'                                    
2022-06-02 21:35:50.582 :: 2022-06-02 21:35:50.383    ::   Preparing metadata (pyproject.toml): started                                                        
2022-06-02 21:36:05.603 :: 2022-06-02 21:35:50.582    ::   Preparing metadata (pyproject.toml): finished with status 'done'                                    
```


i wish i knew what was going on during that "build dependencies" step -- whatever it is requires deep thought

**Comment by facundobatista** [comment]
For corner cases, binary dependencies can be installed. Use the the `charm-binary-python-packages` key in `charmcraft.yaml` (as indicated (in the [config reference](https://juju.is/docs/sdk/charmcraft-config)). 

I mean "corner cases" because we firmly encourage to install packages from sources so all used/deployed code can be tracked.
